testing/unstable_stub_test.ts

```diff
@@ -0,0 +1,394 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+import { assertEquals, assertThrows } from "@std/assert";
+import {
+  assertSpyCall,
+  assertSpyCallArg,
+  assertSpyCalls,
+  MockError,
+  returnsNext,
+} from "./mock.ts";
+import { Point, type PointWithExtra } from "./_test_utils.ts";
+import { type Stub, stub } from "./unstable_stub.ts";
+
+Deno.test("stub()", () => {
+  const point = new Point(2, 3);
+  const func = stub(point, "action");
+
+  assertSpyCalls(func, 0);
+
+  assertEquals(func.call(point), undefined);
+  assertSpyCall(func, 0, {
+    self: point,
+    args: [],
+    returned: undefined,
+  });
+  assertSpyCalls(func, 1);
+
+  assertEquals(point.action(), undefined);
+  assertSpyCall(func, 1, {
+    self: point,
+    args: [],
+    returned: undefined,
+  });
+  assertSpyCalls(func, 2);
+
+  assertEquals(func.original, Point.prototype.action);
+  assertEquals(point.action, func);
+
+  assertEquals(func.restored, false);
+  func.restore();
+  assertEquals(func.restored, true);
+  assertEquals(point.action, Point.prototype.action);
+  assertThrows(
+    () => func.restore(),
+    MockError,
+    "Cannot restore: instance method already restored",
+  );
+  assertEquals(func.restored, true);
+});
+
+Deno.test("stub() works on function", () => {
+  const point = new Point(2, 3);
+  const returns = [1, "b", 2, "d"];
+  const func = stub(point, "action", () => returns.shift());
+
+  assertSpyCalls(func, 0);
+
+  assertEquals(func.call(point), 1);
+  assertSpyCall(func, 0, {
+    self: point,
+    args: [],
+    returned: 1,
+  });
+  assertSpyCalls(func, 1);
+
+  assertEquals(point.action(), "b");
+  assertSpyCall(func, 1, {
+    self: point,
+    args: [],
+    returned: "b",
+  });
+  assertSpyCalls(func, 2);
+
+  assertEquals(func.original, Point.prototype.action);
+  assertEquals(point.action, func);
+
+  assertEquals(func.restored, false);
+  func.restore();
+  assertEquals(func.restored, true);
+  assertEquals(point.action, Point.prototype.action);
+  assertThrows(
+    () => func.restore(),
+    MockError,
+    "Cannot restore: instance method already restored",
+  );
+  assertEquals(func.restored, true);
+});
+Deno.test("stub() works on getter", () => {
```

Comment:
Technically, on the point class, both x and y are values. We are overriding the property descriptor with one that has a getter instead of a value. Same with the other tests. I would describe this, "stub() property value with a getter".

We should get a test case that shows stubbing a property that originally had a getter, setter, or both works too. In all cases, restore should change the property descriptor to what it was before we stubbed it.

Reply:
For that last part, I think the test cases should be updated to first get a copy of the property descriptor for point.x, then at the end after restoring, assert that the property descriptor for point.x matches what it was originally.